File: generateparentheses/generate_parentheses.py
from typing import List

class Solution:

    def _insert_inside(self, string: str, left_index: int) -> str:

        left_part = string[:left_index + 1]
        right_part = string[left_index + 1:]

        return f"{left_part}(){right_part}"

    def generateParenthesis(self, n: int) -> List[str]:

        solutions = set()
        if not n:
            solutions.add("")
        else:
            previous_solution = self.generateParenthesis(n-1)
            for string in previous_solution:
                for ind, _ in enumerate(string):
                    if string[ind] == "(":
                        solution = self._insert_inside(string, ind)
                        solutions.add(solution)
                solutions.add("()" + string)
        
        return list(solutions)


if __name__ == "__main__":
    print(Solution().generateParenthesis(3))

File: generateparentheses/generate_parentheses.go
package generateparentheses

func insertInside(str string, leftIndex int) string {

	leftPart := str[:leftIndex+1]
	rightPart := str[leftIndex+1:]

	return leftPart + "()" + rightPart
}

func generateParenthesis(n int) map[string]int {

	solutions := make(map[string]int)
	if n == 0 {
		solutions[""] = 1
	} else {
		prevSolution := generateParenthesis(n - 1)

		for str := range prevSolution {
			for ind := range str {
				if str[ind] == '(' {
					solution := insertInside(str, ind)
					solutions[solution] = 1
				}
			}
			solutions["()"+str] = 1
		}
	}

	return solutions
}

// GenerateParenthesis ...
func GenerateParenthesis(n int) []string {
	solutions := generateParenthesis(n)
	var finalSolutions []string

	for solution := range solutions {
		if solution != "" {
			finalSolutions = append(finalSolutions, solution)
		}
	}

	return finalSolutions
}
